src/interp/interp-wasi.cc

```diff
@@ -76,18 +120,247 @@ class WasiInstance {
+    return Result::Ok;
+  }
+
+  Result path_rename(const Values& params, Values& results, Trap::Ptr* trap) {
+    assert(false);
+    return Result::Ok;
+  }
+
+  Result path_open(const Values& params, Values& results, Trap::Ptr* trap) {
+    /* __wasi_errno_t __wasi_path_open(__wasi_fd_t fd,
+                                       __wasi_lookupflags_t dirflags,
+                                       const char *path,
+                                       size_t path_len,
+                                       __wasi_oflags_t oflags,
+                                       __wasi_rights_t fs_rights_base,
+                                       __wasi_rights_t fs_rights_inherting,
+                                       __wasi_fdflags_t fdflags,
+                                       __wasi_fd_t *opened_fd) */
+    uvwasi_fd_t dirfd = params[0].i32_;
+    __wasi_lookupflags_t dirflags = params[1].i32_;
+    uint32_t path_ptr = params[2].i32_;
+    __wasi_size_t path_len = params[3].i32_;
+    __wasi_oflags_t oflags = params[4].i32_;
+    __wasi_rights_t fs_rights_base = params[5].i32_;
+    __wasi_rights_t fs_rights_inherting = params[6].i32_;
+    __wasi_fdflags_t fs_flags = params[7].i32_;
+    uint32_t out_ptr = params[8].i32_;
+    char* path;
+    CHECK_RESULT(getMemPtr<char>(path_ptr, path_len, &path, trap));
+    if (trace_stream) {
+      trace_stream->Writef("path_open : %s\n", path);
+    }
+    uvwasi_fd_t outfd;
+    results[0].i32_ =
+        uvwasi_path_open(uvwasi, dirfd, dirflags, path, path_len, oflags,
+                         fs_rights_base, fs_rights_inherting, fs_flags, &outfd);
+    if (trace_stream) {
+      trace_stream->Writef("path_open -> %d\n", results[0].i32_);
+    }
+    CHECK_RESULT(writeValue<__wasi_fd_t>(outfd, out_ptr, trap));
+    return Result::Ok;
+  }
+
+  Result path_filestat_get(const Values& params,
+                           Values& results,
+                           Trap::Ptr* trap) {
+    /* __wasi_errno_t __wasi_path_filestat_get(__wasi_fd_t fd,
+     *                                         __wasi_lookupflags_t flags,
+     *                                         const char *path,
+     *                                         size_t path_len,
+     *                                         __wasi_filestat_t *buf
+     */
+    uvwasi_fd_t fd = params[0].i32_;
+    __wasi_lookupflags_t flags = params[1].i32_;
+    uint32_t path_ptr = params[2].i32_;
+    uvwasi_size_t path_len = params[3].i32_;
+    uint32_t filestat_ptr = params[4].i32_;
+    char* path;
+    CHECK_RESULT(getMemPtr<char>(path_ptr, path_len, &path, trap));
+    if (trace_stream) {
+      trace_stream->Writef("path_filestat_get : %d %s\n", fd, path);
+    }
+    uvwasi_filestat_t buf;
+    results[0].i32_ =
+        uvwasi_path_filestat_get(uvwasi, fd, flags, path, path_len, &buf);
+    uvwasi_serdes_write_filestat_t(memory->UnsafeData(), filestat_ptr, &buf);
```

Comment:
Check that this is a valid range first? Or is that tricky since it's not known what the serialized size is?

Reply:
Updated all these to use getMemPtr.